### PR TITLE
8322282: Incorrect LoaderConstraintTable::add_entry after JDK-8298468

### PR DESCRIPTION
**Pull request by shipilev**
As reported here:
https://mail.openjdk.org/pipermail/hotspot-compiler-dev/2023-December/071086.html

There is a regression introduced by [JDK-8298468](https://bugs.openjdk.org/browse/JDK-8298468), which looks like a simple typo:
https://github.com/openjdk/jdk/commit/10737e168c967a08e257927251861bf2c14795ab#diff-49d18b7f77db80ff67872c7db1e4c87e98a126e63fff5f63353e4406db332b19R343

Additional testing:
 - [x] Linux x86_64 server fastdebug,`tier{1,2,3}`
 - [x] Linux AArch64 server fastdebug,`tier{1,2,3}`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322282](https://bugs.openjdk.org/browse/JDK-8322282): Incorrect LoaderConstraintTable::add_entry after JDK-8298468 (**Bug** - P2)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [1d73bb64](https://git.openjdk.org/jdk/pull/17140/files/1d73bb6456fdef857aaafae45b70a6b535ed927c)
 * [Volker Simonis](https://openjdk.org/census#simonis) (@simonis - **Reviewer**) ⚠️ Review applies to [1d73bb64](https://git.openjdk.org/jdk/pull/17140/files/1d73bb6456fdef857aaafae45b70a6b535ed927c)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [1d73bb64](https://git.openjdk.org/jdk/pull/17140/files/1d73bb6456fdef857aaafae45b70a6b535ed927c)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17140/head:pull/17140` \
`$ git checkout pull/17140`

Update a local copy of the PR: \
`$ git checkout pull/17140` \
`$ git pull https://git.openjdk.org/jdk.git pull/17140/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17140`

View PR using the GUI difftool: \
`$ git pr show -t 17140`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17140.diff">https://git.openjdk.org/jdk/pull/17140.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17140#issuecomment-1860319585)

### CODE REVIEW

**Review by dholmes-ora — COMMENTED** [review]
I will need to check with the original fix but regardless this is obviously crying out for a regression test.

I and I still can't see any way that this leads to the NPE that was reported.

**Comment by shipilev** [comment]
> I will need to check with the original fix

Look at the change that introduced it: https://github.com/openjdk/jdk/commit/10737e168c967a08e257927251861bf2c14795ab?diff=unified&w=0#diff-49d18b7f77db80ff67872c7db1e4c87e98a126e63fff5f63353e4406db332b19R340-R344

...it looks like a typo, isn't it? The logic for that code is also understandable: if one `LoaderConstraint` is `nullptr`, we merge into another one.

> I and I still can't see any way that this leads to the NPE that was reported.

I see compiler looks up classes from `LoaderConstraintTable` sometimes: `ciEnv::get_klass_by_name_impl` -> `SystemDictionary::find_constrained_instance_or_array_klass`. So it makes some sense to me that if we are reporting wrong result to compiler, it would then foobar the generated code by assuming something about the class that is not true.

**Comment by shipilev** [comment]
> this is obviously crying out for a regression test.

Yes, maybe. I would prefer to get the trivial fix that is known to affect real-world JDK 21 use without stalling on creating, debugging, reviewing the regression test.

**Comment by dholmes-ora** [comment]
Problem is without a proper regression test we only know that this change fixes one problem, but we don't know if it has any unexpected side-effects that the bug has been masking. I want a much better understanding of the bug before integrating the fix.

**Comment by shipilev** [comment]
> Problem is without a proper regression test we only know that this change fixes one problem, but we don't know if it has any unexpected side-effects that the bug has been masking. I want a much better understanding of the bug before integrating the fix.

From the strategical point of view, I understand the desire to understand the compile (mis)behavior better. But I think it is not a great tactical move, as it is not directly relevant to the fix.

Here is how I look at this problem:
 1. [JDK-8298468](https://bugs.openjdk.org/browse/JDK-8298468) was supposed to be a cleanup. In other words, it is was supposed to be behavior-preserving.
 2. We have identified the hunk in the original changeset that violates the behavior-preserving property, which looks like a simple mistake.
 3. This fix restores the original behavior before [JDK-8298468](https://bugs.openjdk.org/browse/JDK-8298468) in this hunk.

The good questions to ask ourselves at this point are: _Would any of these facts change as the result of the deeper understanding of how exactly this mistake breaks the compiler? Would there be a situation where we decide that going with this fix is not a good idea? Would we be deeply investigating the implications of this mistake if it was caught during code review, instead of just fixing it and moving on?_ I think the probability of answering "yes" to all these questions is very low. In fact the only way I see that happening if we discover that the whole [JDK-8298468](https://bugs.openjdk.org/browse/JDK-8298468) is wrong and needs to be backed out.

But if we (I argue, reasonably) answer "no" to these questions, then holding off from integrating this fix before we gain more understanding how exactly it breaks compiler gains us nothing. But it delays fixing JDK 23, 23, 21 sooner, which is even more important when there is an actual _customer-visible breakage_ reported on JDK 21.

In other words, could there be a situation where this mistake exposes some deeper problem in Hotspot? Yes, that might happen, but it changes nothing about the actual behavior-breaking impact of [JDK-8298468](https://bugs.openjdk.org/browse/JDK-8298468) this point fix resolves. One can always follow-up deeper on these problems by reverting the fix in mainline and then investigating. I would further suspect that this is a kind of "garbage in, garbage out" scenario, where we report wrong metadata to the compiler, and it breaks in fascinating ways. We know it happens in other funky scenarios like feeding the unverifiable bytecode, for example.

So, I propose we do not quibble over this fix.

**Comment by shipilev** [comment]
Not now, bot.

**Comment by simonis** [comment]
@coleenp can you please take a quick look at this fix which is a regression of [JDK-8298468](https://bugs.openjdk.org/browse/JDK-8298468).

**Review by dholmes-ora — APPROVED** [review]
We will continue to think about how to produce a regression test for this, but the fix is cleared for integration.

Thanks

**Review by coleenp — APPROVED** [review]
Yes, this looks correct.  Thanks to Antoine for reporting it and your diagnosis.

**Review by coleenp — COMMENTED** [review]
You probably should update the copyright to 2024.

**Comment by adessaigne** [comment]
> Yes, this looks correct. Thanks to Antoine for reporting it and your diagnosis.

No problem. I learned a lot while diagnosing it. Thank you all for your help in this matter.

**Comment by shipilev** [comment]
Thank you all! I merged from master, fixed the copyright year, and would integrate as soon as light testing turns green.

**Comment by shipilev** [comment]
Testing is clean.

/integrate
